src/utils/saga/paged.js

```diff
@@ -0,0 +1,123 @@
+import { call } from 'redux-saga/effects';
+
+import request from 'utils/request';
+
+/**
+ * Counter Helper
+ * @param {url} url Route to the counter
+ * @param {type} type What is being counted
+ * @eample { url: '/roles/count', type: 'Roles' }
+ */
+export function* counter({ url, type }) {
+  const endpoint = {
+    url,
+    method: 'GET'
+  };
+  const count = yield call(request.execute, { endpoint });
+
+  if (count.success) {
+    const {
+      response: {
+        data: { result }
+      }
+    } = count;
+
+    return result;
+  } else if (count.error) {
+    throw count.error;
+  } else {
+    throw new Error(`Failed to count ${type}!`);
+  }
+}
+
+/**
+ * Pager Helper
+ * @param {req} req
+ * @example { cached: rolesCached,
+  pager: {
+    ...pager,
+    store: rolesPager
+  },
+  path: {
+    counter: '/roles/count',
+    query: '/roles/'
+  },
+  type: 'Roles' }
+ */
+export function* pager({ req }) {
+  const Pager = {};
+
+  const {
+    pager: { limit, page, store },
+    path,
+    type
+  } = req;
+
+  Pager.count = !store.count
+    ? yield call(counter, { url: path.counter, type })
+    : store.count;
+  Pager.limit = !limit ? store.limit : limit;
+  Pager.pages =
+    !store.pages || store.pages === null || limit !== store.limit
+      ? Math.ceil(Pager.count / Pager.limit)
+      : store.pages;
+  Pager.page = !page ? store.page : page;
+
+  return Pager;
+}
+
+/**
+ * Paged Helper
+ * @param {*} req
+ * @example { cached: rolesCached,
+  pager: {
+    ...pager,
+    store: rolesPager
+  },
+  path: {
+    counter: '/roles/count',
+    query: '/roles/'
+  },
+  type: 'Roles' }
+ */
+export function* paged(req) {
+  const Pager = yield call(pager, { req });
+  const { count, limit, page } = Pager;
+  const { path, cached, type } = req;
+
+  if (
+    !cached[page] ||
+    (count > Number(limit) && cached[page].length !== Number(limit)) ||
+    (count < Number(limit) && cached[page].length !== count)
+  ) {
+    let offset = page * limit - limit + 1;
+
+    if (offset > count) {
+      offset = count - limit;
+    }
+    const endpoint = {
+      url: `${path.query}?limit=${limit}&offset=${offset}`,
```

Comment:
Will we ever need to _append_ to an existing query string? i.e. if `path.query` is already `/some/call?withparams=likethis`

Reply:
I needed to remove confusion here, so I renamed path to route and path.query to route.data. It is just the route to call (/roles for example), the parameters are appended after the pager determines offset.